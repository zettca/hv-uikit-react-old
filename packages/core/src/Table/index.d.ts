export * from "../TableNew";
export * from "../TableOld";
