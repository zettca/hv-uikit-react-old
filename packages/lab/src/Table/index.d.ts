export { default } from "./Table";
export * from "./Table";

export { default as HvTableBody } from "./TableBody";
export * from "./TableBody";

export { default as HvTableCell } from "./TableCell";
export * from "./TableCell";

export { default as HvTableContainer } from "./TableContainer";
export * from "./TableContainer";

export { default as HvTableHead } from "./TableHead";
export * from "./TableHead";

export { default as HvTablePagination } from "./TablePagination";
export * from "./TablePagination";

export { default as HvTableRow } from "./TableRow";
export * from "./TableRow";
