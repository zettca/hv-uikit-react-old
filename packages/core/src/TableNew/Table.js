import React from "react";
import PropTypes from "prop-types";
import clsx from "clsx";

import { Table, withStyles } from "@material-ui/core";

import styles from "./styles";

/**
 * Table description/documentation paragraph
 */
const HvTable = (props) => {
  const { className, classes, children, ...others } = props;

  return (
    <Table size="small" className={clsx(className, classes.root)} {...others}>
      {children}
    </Table>
  );
};

HvTable.propTypes = {
  /**
   * Class names to be applied.
   */
  className: PropTypes.string,
  /**
   * Content to be rendered
   */
  children: PropTypes.node.isRequired,
  /**
   * A Jss Object used to override or extend the styles applied.
   */
  classes: PropTypes.shape({
    /**
     * Styles applied to the component root class.
     */
    root: PropTypes.string,
  }).isRequired,
};

export default withStyles(styles, { name: "HvTable" })(HvTable);
