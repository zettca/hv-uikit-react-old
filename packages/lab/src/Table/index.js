export { default } from "./Table";
export { default as HvTableBody } from "./TableBody";
export { default as HvTableCell } from "./TableCell";
export { default as HvTableContainer } from "./TableContainer";
export { default as HvTableHead } from "./TableHead";
export { default as HvTablePagination } from "./TablePagination";
export { default as HvTableRow } from "./TableRow";
