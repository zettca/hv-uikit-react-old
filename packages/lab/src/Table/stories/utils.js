import { useCallback, useMemo, useRef, useState } from "react";
import range from "lodash/range";
import { Random } from "@hv/uikit-react-core";

const rand = new Random();

const formatDate = (date) => date.toISOString().split("T")[0];

const getOption = (opts, i) => opts[i % opts.length];

const newEntry = (i) => {
  const r = rand.next();
  const [dateMax, dateMin] = [2018, 2022].map((y) => new Date(y, 0).getTime());
  return {
    id: `${i + 1}`,
    name: `Event ${i + 1}`,
    createdDate: formatDate(new Date(rand.next(dateMax, dateMin))),
    eventType: "Anomaly detection",
    status: getOption(["Closed", "Open"], i),
    riskScore: rand.next(100, 10),
    severity: getOption(["Critical", "Major", "Average", "Minor"], i),
    priority: (r > 0.66 && "High") || (r > 0.33 && "Medium") || "Low",
  };
};

export const makeData = (len = 10) => range(len).map(newEntry);

// https://react-table.tanstack.com/docs/api/useTable#column-options
// width is only used if explicitly passed in column.getHeaderProps
export const getColumns = () => [
  { Header: "Title", accessor: "name", width: 260 },
  { Header: "Time", accessor: "createdDate" },
  { Header: "Event Type", accessor: "eventType", width: 200 },
  { Header: "Status", accessor: "status", width: 120 },
  // numeric values should be right-aligned
  {
    Header: "Probability",
    accessor: "riskScore",
    align: "right",
    Cell: ({ value }) => `${value}%`,
  },
  { Header: "Severity", accessor: "severity" },
  { Header: "Priority", accessor: "priority" },
];

export const useToggleIndex = (initialState) => {
  const [index, setIndex] = useState(initialState);

  const toggleState = (idx) => () => {
    setIndex(idx === index ? -1 : idx);
  };

  return [index, toggleState];
};

const simpleSortBy = (a, b, sortBy) => {
  const { id, desc } = sortBy;
  if (a[id] < b[id]) return desc ? -1 : 1;
  if (a[id] > b[id]) return desc ? 1 : -1;
  return 0;
};

export const useServerData = () => {
  const serverData = useMemo(() => makeData(999), []);
  const columns = useMemo(() => getColumns().map((col) => ({ ...col, isSortable: true })), []);

  const [data, setData] = useState([]);
  const [loading, setLoading] = useState(false);
  const [pageCount, setPageCount] = useState(0);
  const fetchIdRef = useRef(0);

  const fetchData = useCallback(
    ({ pageSize, pageIndex, sortBy }) => {
      fetchIdRef.current += 1;
      const fetchId = fetchIdRef.current;

      setLoading(true);

      setTimeout(() => {
        if (fetchId !== fetchIdRef.current) return;

        const startRow = pageSize * pageIndex;
        const endRow = startRow + pageSize;

        const newData = [...serverData]
          .sort(sortBy.length === 0 ? undefined : (a, b) => simpleSortBy(a, b, sortBy[0]))
          .slice(startRow, endRow);
        setData(newData);

        // Static example . Set server-side page count here
        setPageCount(Math.ceil(serverData.length / pageSize));

        setLoading(false);
      }, 600);
    },
    [serverData]
  );

  return [data, columns, fetchData, loading, pageCount];
};
