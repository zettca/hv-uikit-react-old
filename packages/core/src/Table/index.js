import React from "react";
import TableNew from "../TableNew";
import TableOld from "../TableOld";

const TableSwitcher = (props) => {
  // eslint-disable-next-line react/prop-types
  const { children } = props;
  return children ? <TableNew {...props} /> : <TableOld {...props} />;
};

export default TableSwitcher;

export * from "../TableNew";
