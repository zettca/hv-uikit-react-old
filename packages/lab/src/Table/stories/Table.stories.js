import React, { useEffect, useMemo } from "react";
import range from "lodash/range";
import { useTable, useRowSelect, usePagination, useSortBy, useResizeColumns } from "react-table";

import { Ban } from "@hv/uikit-react-icons";

import { HvCheckBox, HvDropDownMenu, HvEmptyState, HvLoading } from "@hv/uikit-react-core";

import {
  HvTable,
  HvTableBody,
  HvTableCell,
  HvTableContainer,
  HvTableHead,
  HvTablePagination,
  HvTableRow,
} from "../..";

import { makeData, getColumns, useToggleIndex, useServerData } from "./utils";

export default {
  title: "Lab/Table",
  parameters: {
    componentSubtitle: null,
    usage: "import { HvTable } from '@hv/uikit-react-core'",
  },
  component: HvTable,
  subcomponents: {
    HvTableBody,
    HvTableCell,
    HvTableContainer,
    HvTableHead,
    HvTablePagination,
    HvTableRow,
  },
};

export const Main = () => (
  <HvTableContainer>
    <HvTable>
      <HvTableHead>
        <HvTableRow>
          {getColumns().map((el) => (
            <HvTableCell key={el.Header}>{el.Header}</HvTableCell>
          ))}
        </HvTableRow>
      </HvTableHead>
      <HvTableBody>
        {makeData(6).map((el) => (
          <HvTableRow key={el.id} hover>
            <HvTableCell>{el.name}</HvTableCell>
            <HvTableCell>{el.createdDate}</HvTableCell>
            <HvTableCell>{el.eventType}</HvTableCell>
            <HvTableCell>{el.status}</HvTableCell>
            <HvTableCell>{el.riskScore}</HvTableCell>
            <HvTableCell>{el.severity}</HvTableCell>
            <HvTableCell>{el.priority}</HvTableCell>
          </HvTableRow>
        ))}
      </HvTableBody>
    </HvTable>
  </HvTableContainer>
);

Main.parameters = {
  docs: {
    description: { story: "A minimal table example." },
  },
};

export const Empty = () => {
  const EmptyRow = () => (
    <HvTableRow>
      <HvTableCell colSpan="100%" style={{ height: 100 }}>
        <HvEmptyState message="No data to display" icon={<Ban role="presentation" />} />
      </HvTableCell>
    </HvTableRow>
  );
  return (
    <HvTableContainer>
      <HvTable>
        <HvTableHead>
          <HvTableRow>
            {getColumns().map((el) => (
              <HvTableCell key={el.Header}>{el.Header}</HvTableCell>
            ))}
          </HvTableRow>
        </HvTableHead>
        <HvTableBody>
          <EmptyRow />
        </HvTableBody>
      </HvTable>
    </HvTableContainer>
  );
};

Main.parameters = {
  docs: {
    description: { story: "A minimal table example." },
  },
};

export const SelectableSimple = () => {
  const [checkedIdx, toggleChecked] = useToggleIndex(0);
  const data = useMemo(() => makeData(6), []);
  const actions = useMemo(() => range(3).map((i) => ({ label: `Option ${i + 1}` })), []);

  return (
    <HvTableContainer>
      <HvTable>
        <HvTableHead>
          <HvTableRow>
            <HvTableCell padding="checkbox" />
            {getColumns().map((el) => (
              <HvTableCell key={el.Header}>{el.Header}</HvTableCell>
            ))}
            <HvTableCell padding="checkbox" />
          </HvTableRow>
        </HvTableHead>
        <HvTableBody>
          {data.map((el, idx) => (
            <HvTableRow key={el.id} hover selected={checkedIdx === idx}>
              <HvTableCell padding="checkbox">
                <HvCheckBox checked={checkedIdx === idx} onClick={toggleChecked(idx)} />
              </HvTableCell>
              <HvTableCell>{el.name}</HvTableCell>
              <HvTableCell>{el.createdDate}</HvTableCell>
              <HvTableCell>{el.eventType}</HvTableCell>
              <HvTableCell>{el.status}</HvTableCell>
              <HvTableCell>{el.riskScore}</HvTableCell>
              <HvTableCell>{el.severity}</HvTableCell>
              <HvTableCell>{el.priority}</HvTableCell>
              <HvTableCell padding="checkbox">
                <HvDropDownMenu keepOpened={false} placement="left" dataList={actions} />
              </HvTableCell>
            </HvTableRow>
          ))}
        </HvTableBody>
      </HvTable>
    </HvTableContainer>
  );
};

SelectableSimple.parameters = {
  docs: {
    description: { story: "A table with checkboxes being managed by a simple hook." },
  },
};

export const SelectableReactTable = () => {
  const data = useMemo(() => makeData(6), []);
  const columns = useMemo(
    () => [
      {
        id: "selection",
        padding: "checkbox",
        Cell: ({ row }) => <HvCheckBox {...row.getToggleRowSelectedProps()} />,
      },
      ...getColumns(),
    ],
    []
  );

  const instance = useTable({ columns, data }, useRowSelect);
  const { getTableProps, getTableBodyProps, prepareRow, headers, rows } = instance;

  return (
    <HvTableContainer>
      <HvTable {...getTableProps()}>
        <HvTableHead>
          <HvTableRow>
            {headers.map((col) => (
              <HvTableCell key={col.Header} rtCol={col} {...col.getHeaderProps()}>
                {col.render("Header")}
              </HvTableCell>
            ))}
          </HvTableRow>
        </HvTableHead>
        <HvTableBody {...getTableBodyProps()}>
          {rows.map((row) => {
            prepareRow(row);
            return (
              <HvTableRow hover key={row.Header} selected={row.isSelected} {...row.getRowProps()}>
                {row.cells.map((cell) => (
                  <HvTableCell key={cell.Header} rtCol={cell.column} {...cell.getCellProps()}>
                    {cell.render("Cell")}
                  </HvTableCell>
                ))}
              </HvTableRow>
            );
          })}
        </HvTableBody>
      </HvTable>
    </HvTableContainer>
  );
};

SelectableReactTable.parameters = {
  docs: {
    description: { story: "A table with checkboxes being managed by `react-table`." },
  },
};

export const Sortable = () => {
  const data = useMemo(() => makeData(5), []);
  const columns = useMemo(() => getColumns().map((col) => ({ ...col, isSortable: true })), []);

  const instance = useTable({ columns, data }, useSortBy, useRowSelect);
  const { getTableProps, getTableBodyProps, headers, rows, prepareRow } = instance;

  return (
    <HvTableContainer>
      <HvTable {...getTableProps()}>
        <HvTableHead>
          <HvTableRow>
            {headers.map((col) => (
              <HvTableCell
                key={col.Header}
                rtCol={col}
                {...col.getHeaderProps(col.getSortByToggleProps({ style: { width: col.width } }))}
              >
                {col.render("Header")}
              </HvTableCell>
            ))}
          </HvTableRow>
        </HvTableHead>
        <HvTableBody {...getTableBodyProps()}>
          {rows.map((row) => {
            prepareRow(row);
            return (
              <HvTableRow hover {...row.getRowProps()}>
                {row.cells.map((cell) => (
                  <HvTableCell rtCol={cell.column} {...cell.getCellProps()}>
                    {cell.render("Cell")}
                  </HvTableCell>
                ))}
              </HvTableRow>
            );
          })}
        </HvTableBody>
      </HvTable>
    </HvTableContainer>
  );
};

Sortable.parameters = {
  docs: {
    description: { story: "A table with multi column sorting, managed by `react-table`." },
  },
};

export const Pagination = () => {
  const data = useMemo(() => makeData(32), []);
  const columns = useMemo(() => getColumns(), []);

  const instance = useTable({ columns, data }, usePagination, useResizeColumns);
  const {
    getTableProps,
    getTableBodyProps,
    prepareRow,
    headers,
    page,
    state: { pageSize },
  } = instance;

  const EmptyRow = () => (
    <HvTableRow>
      <HvTableCell colSpan="100%" />
    </HvTableRow>
  );

  return (
    <div>
      <HvTableContainer>
        <HvTable {...getTableProps()}>
          <HvTableHead>
            <HvTableRow>
              {headers.map((col) => (
                <HvTableCell
                  key={col.Header}
                  rtCol={col}
                  // adds fixed column sizes
                  {...col.getHeaderProps({ style: { width: col.width } })}
                >
                  {col.render("Header")}
                </HvTableCell>
              ))}
            </HvTableRow>
          </HvTableHead>
          <HvTableBody {...getTableBodyProps()}>
            {range(pageSize).map((i) => {
              const row = page[i];

              if (!row) return <EmptyRow key={i} />;

              prepareRow(row);
              return (
                <HvTableRow key={i} {...row.getRowProps()}>
                  {row.cells.map((cell) => (
                    <HvTableCell key={cell.Header} rtCol={cell.column} {...cell.getCellProps()}>
                      {cell.render("Cell")}
                    </HvTableCell>
                  ))}
                </HvTableRow>
              );
            })}
          </HvTableBody>
        </HvTable>
      </HvTableContainer>
      <HvTablePagination rtInstance={instance} />
    </div>
  );
};

Pagination.parameters = {
  docs: {
    description: { story: "A table with pagination managed by `react-table`." },
  },
};

export const ServerSide = () => {
  const [data, columns, fetchData, loading, pageCount] = useServerData();

  const instance = useTable(
    {
      columns,
      data,
      manualPagination: true,
      manualSortBy: true,
      autoResetPage: false,
      autoResetSortBy: false,
      pageCount,
    },
    useSortBy,
    usePagination,
    useRowSelect
  );
  const {
    getTableProps,
    getTableBodyProps,
    headers,
    prepareRow,
    page,
    gotoPage,
    state: { pageSize, pageIndex, sortBy },
  } = instance;

  useEffect(() => {
    gotoPage(0);
  }, [sortBy, gotoPage]);

  useEffect(() => {
    fetchData({ pageIndex, pageSize, sortBy });
  }, [sortBy, fetchData, pageIndex, pageSize]);

  const EmptyRow = () => (
    <HvTableRow>
      <HvTableCell colSpan="100%" />
    </HvTableRow>
  );

  return (
    <div>
      {loading && (
        <div style={{ position: "absolute", height: 0, left: 0, right: 0, top: 80 }}>
          <HvLoading />
        </div>
      )}
      <HvTableContainer>
        <HvTable {...getTableProps()}>
          <HvTableHead>
            <HvTableRow>
              {headers.map((col) => (
                <HvTableCell
                  key={col.Header}
                  rtCol={col}
                  {...col.getHeaderProps(col.getSortByToggleProps({ style: { width: col.width } }))}
                >
                  {col.render("Header")}
                </HvTableCell>
              ))}
            </HvTableRow>
          </HvTableHead>
          <HvTableBody {...getTableBodyProps({ style: { position: "relative" } })}>
            {range(pageSize).map((i) => {
              const row = page[i];

              if (!row) return <EmptyRow key={i} />;

              prepareRow(row);
              return (
                <HvTableRow hover {...row.getRowProps()}>
                  {row.cells.map((cell) => (
                    <HvTableCell rtCol={cell.column} {...cell.getCellProps()}>
                      {cell.render("Cell")}
                    </HvTableCell>
                  ))}
                </HvTableRow>
              );
            })}
          </HvTableBody>
        </HvTable>
      </HvTableContainer>
      <HvTablePagination rtInstance={instance} />
    </div>
  );
};

ServerSide.parameters = {
  docs: {
    description: {
      story:
        "A table with sorting and pagination handled server-side, using React Table. Set `manualPagination` and `manualSortBy` to have manual control over pagination and sorting.",
    },
  },
};
