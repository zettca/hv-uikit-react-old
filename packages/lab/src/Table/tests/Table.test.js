/* eslint-env jest */

import React from "react";
import range from "lodash/range";
import { render } from "testing-utils";
import userEvent from "@testing-library/user-event";

import { Main, Pagination, Sortable } from "../stories/Table.stories";
import {
  HvTable,
  HvTableBody,
  HvTableCell,
  HvTableContainer,
  HvTableHead,
  HvTableRow,
} from "../..";

describe("Table", () => {
  describe("Main Story", () => {
    it("should match snapshot", () => {
      const { container } = render(<Main />);
      expect(container).toBeDefined();
      expect(container).toMatchSnapshot();
    });

    it("should render the table elements", () => {
      const { getByRole, getAllByRole } = render(<Main />);

      expect(getByRole("table")).toBeInTheDocument();

      expect(getAllByRole("cell").length).toBeGreaterThan(0);
      expect(getAllByRole("row").length).toBeGreaterThan(0);
    });
  });

  describe("Simple Table", () => {
    const NUM_ROWS = 6;
    const NUM_COLS = 3;

    const SimpleTable = () => (
      <HvTableContainer>
        <HvTable>
          <HvTableHead>
            <HvTableRow>
              {range(NUM_COLS).map((id) => (
                <HvTableCell key={id}>{`Sample Header ${id}`}</HvTableCell>
              ))}
            </HvTableRow>
          </HvTableHead>
          <HvTableBody>
            {range(NUM_ROWS).map((id) => (
              <HvTableRow key={id}>
                {range(NUM_COLS).map((id2) => (
                  <HvTableCell key={id2}>{`Sample Cell ${id2}`}</HvTableCell>
                ))}
              </HvTableRow>
            ))}
          </HvTableBody>
        </HvTable>
      </HvTableContainer>
    );

    it("should be defined", () => {
      const { container } = render(<SimpleTable />);

      expect(container).toBeDefined();
    });

    it("should render the rows and cells", () => {
      const { getAllByRole } = render(<SimpleTable />);

      expect(getAllByRole("rowgroup").length).toBe(2); // thead & tbody
      expect(getAllByRole("row").length).toBe(NUM_ROWS + 1);
      expect(getAllByRole("columnheader").length).toBe(NUM_COLS);
      expect(getAllByRole("cell").length).toBe(NUM_COLS * NUM_ROWS);
    });
  });

  describe("Header Only Table", () => {
    const NUM_COLS = 6;

    const HeadersOnlyTable = () => (
      <HvTableContainer>
        <HvTable>
          <HvTableHead>
            <HvTableRow>
              {range(NUM_COLS).map((id) => (
                <HvTableCell key={id}>{`Header ${id}`}</HvTableCell>
              ))}
            </HvTableRow>
          </HvTableHead>
          <HvTableBody />
        </HvTable>
      </HvTableContainer>
    );

    it("should render single row and its cells", () => {
      const { getAllByRole } = render(<HeadersOnlyTable />);

      expect(getAllByRole("row").length).toBe(1);
      expect(getAllByRole("columnheader").length).toBe(NUM_COLS);
    });
  });

  describe("Pagination Story", () => {
    it("should be defined", () => {
      const { container } = render(<Pagination />);
      expect(container).toBeDefined();
    });

    it("should render the table elements", () => {
      const { getByRole, getAllByRole } = render(<Pagination />);

      expect(getByRole("table")).toBeInTheDocument();

      expect(getAllByRole("cell").length).toBeGreaterThan(0);
      expect(getAllByRole("row").length).toBeGreaterThan(0);
    });

    it("should contain the correct page elements", () => {
      const { getByLabelText, queryByText, getAllByRole } = render(<Pagination />);

      const [fistPage, previousPage, nextPage, lastPage] = [
        "First Page",
        "Previous Page",
        "Next Page",
        "Last Page",
      ].map(getByLabelText);

      expect(getAllByRole("row").length).toBe(11);
      expect(queryByText("Event 1")).toBeInTheDocument();
      expect(queryByText("Event 11")).not.toBeInTheDocument();

      userEvent.click(fistPage);
      expect(getAllByRole("row").length).toBe(11);
      expect(queryByText("Event 1")).toBeInTheDocument();
      expect(queryByText("Event 11")).not.toBeInTheDocument();

      userEvent.click(previousPage);
      expect(getAllByRole("row").length).toBe(11);
      expect(queryByText("Event 1")).toBeInTheDocument();
      expect(queryByText("Event 11")).not.toBeInTheDocument();

      userEvent.click(nextPage);
      expect(getAllByRole("row").length).toBe(11);
      expect(queryByText("Event 1")).not.toBeInTheDocument();
      expect(queryByText("Event 11")).toBeInTheDocument();

      userEvent.click(lastPage);
      expect(getAllByRole("row").length).toBe(11);
      expect(queryByText("Event 1")).not.toBeInTheDocument();
      expect(queryByText("Event 30")).not.toBeInTheDocument();
      expect(queryByText("Event 31")).toBeInTheDocument();
    });
  });

  describe("Sortable Story", () => {
    it("should be defined", () => {
      const { container } = render(<Sortable />);
      expect(container).toBeDefined();
    });

    it("should render the table elements", () => {
      const { getByRole, getAllByRole } = render(<Sortable />);

      expect(getByRole("table")).toBeInTheDocument();

      expect(getAllByRole("cell").length).toBeGreaterThan(0);
      expect(getAllByRole("row").length).toBeGreaterThan(0);
    });

    it("should single-sort as expected", () => {
      const { getByText, queryAllByText } = render(<Sortable />);
      const probabilityButton = getByText("Probability");

      const getEvent = (i) => queryAllByText(/^Event \d+$/g)[i];

      // Default sorting
      expect(getEvent(0)).toHaveTextContent("Event 1");

      // Asc sorting
      userEvent.click(probabilityButton);
      expect(getEvent(0)).toHaveTextContent("Event 3");

      // Desc sorting
      userEvent.click(probabilityButton);
      expect(getEvent(0)).toHaveTextContent("Event 5");

      // Back to default sorting
      userEvent.click(probabilityButton);
      expect(getEvent(0)).toHaveTextContent("Event 1");
    });

    it("should multi-sort as expected", () => {
      const { getByText, queryAllByText } = render(<Sortable />);
      const titleButton = getByText("Title");
      const priorityButton = getByText("Priority");
      const probabilityButton = getByText("Probability");

      const getEvent = (i) => queryAllByText(/^Event \d+/g)[i];
      const getSeverity = (i) => queryAllByText(/^(Critical|Major|Average|Minor)$/g)[i];
      const getPriority = (i) => queryAllByText(/^(High|Low)$/g)[i];

      // Default sorting
      expect(getEvent(0)).toHaveTextContent("Event 1");
      expect(getSeverity(0)).toHaveTextContent("Critical");
      expect(getPriority(0)).toHaveTextContent("Low");

      // Asc sorting priority
      userEvent.click(priorityButton);
      expect(getEvent(0)).toHaveTextContent("Event 5");
      expect(getSeverity(0)).toHaveTextContent("Critical");
      expect(getPriority(0)).toHaveTextContent("High");

      // Desc sorting priority
      userEvent.click(priorityButton);
      expect(getEvent(0)).toHaveTextContent("Event 4");
      expect(getSeverity(0)).toHaveTextContent("Minor");
      expect(getPriority(0)).toHaveTextContent("Low");

      // Asc sorting probability
      userEvent.click(probabilityButton, { shiftKey: true });
      expect(getEvent(0)).toHaveTextContent("Event 2");
      expect(getSeverity(0)).toHaveTextContent("Major");
      expect(getPriority(0)).toHaveTextContent("Low");

      // Desc sorting probability
      userEvent.click(probabilityButton, { shiftKey: true });
      expect(getEvent(0)).toHaveTextContent("Event 1");
      expect(getSeverity(0)).toHaveTextContent("Critical");
      expect(getPriority(0)).toHaveTextContent("Low");

      // Single sort reset
      userEvent.click(titleButton);
      expect(getEvent(0)).toHaveTextContent("Event 1");
      expect(getSeverity(0)).toHaveTextContent("Critical");
      expect(getPriority(0)).toHaveTextContent("Low");
    });
  });
});
