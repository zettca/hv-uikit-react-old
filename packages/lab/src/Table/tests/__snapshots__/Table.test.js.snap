// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Main Story should match snapshot 1`] = `
<div>
  <div
    class="HvTableContainer-root"
  >
    <table
      class="HvTable-root"
    >
      <thead
        class="HvTableHead-root"
      >
        <tr
          class="HvTableRow-root HvTableRow-head"
        >
          <th
            class="HvTableCell-root HvTableCell-head"
          >
            Title
          </th>
          <th
            class="HvTableCell-root HvTableCell-head"
          >
            Time
          </th>
          <th
            class="HvTableCell-root HvTableCell-head"
          >
            Event Type
          </th>
          <th
            class="HvTableCell-root HvTableCell-head"
          >
            Status
          </th>
          <th
            class="HvTableCell-root HvTableCell-head"
          >
            Probability
          </th>
          <th
            class="HvTableCell-root HvTableCell-head"
          >
            Severity
          </th>
          <th
            class="HvTableCell-root HvTableCell-head"
          >
            Priority
          </th>
        </tr>
      </thead>
      <tbody
        class="HvTableBody-root"
      >
        <tr
          class="HvTableRow-root HvTableRow-hover HvTableRow-body"
        >
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Event 1
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            2019-02-28
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Anomaly detection
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Closed
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            98
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Critical
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Low
          </td>
        </tr>
        <tr
          class="HvTableRow-root HvTableRow-hover HvTableRow-body"
        >
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Event 2
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            2021-11-25
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Anomaly detection
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Open
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            32
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Major
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Low
          </td>
        </tr>
        <tr
          class="HvTableRow-root HvTableRow-hover HvTableRow-body"
        >
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Event 3
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            2018-07-15
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Anomaly detection
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Closed
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            63
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Average
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Low
          </td>
        </tr>
        <tr
          class="HvTableRow-root HvTableRow-hover HvTableRow-body"
        >
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Event 4
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            2021-02-26
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Anomaly detection
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Open
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            92
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Minor
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Low
          </td>
        </tr>
        <tr
          class="HvTableRow-root HvTableRow-hover HvTableRow-body"
        >
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Event 5
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            2019-04-27
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Anomaly detection
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Closed
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            16
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Critical
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            High
          </td>
        </tr>
        <tr
          class="HvTableRow-root HvTableRow-hover HvTableRow-body"
        >
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Event 6
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            2021-11-13
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Anomaly detection
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Open
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            98
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            Major
          </td>
          <td
            class="HvTableCell-root HvTableCell-body"
          >
            High
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
